File: 2020/02/day02.R
library(stringr)

puzzle_input <- read.delim("~/Documents/AoC/2020/2/puzzle_input.txt", header = FALSE, sep=":")


## PART 1

true_ct <- 0
false_ct <- 0

for (i in 1:nrow(puzzle_input)){
  # split strings to get min, max, and letter
  min <- as.integer(unlist(strsplit(puzzle_input[i,"V1"], split = "-"))[1])
  max <- as.integer(unlist(strsplit(puzzle_input[i,"V1"], split = "[- ]"))[2])
  letter <- unlist(strsplit(puzzle_input[i,"V1"], split = "[ ]"))[2]
  pw <- puzzle_input[i,"V2"]
  
  
  # test each row for conditions
  if ( str_count(pw, letter) >= min & str_count(pw, letter) <= max){
    true_ct <- true_ct + 1
  } else {
    false_ct <- false_ct + 1
  }
}

print(true_ct)
print(false_ct)

## PART 2

true_ct <- 0
false_ct <- 0

for (i in 1:nrow(puzzle_input)){
  # split strings to get min, max, and letter
  pos1 <- as.integer(unlist(strsplit(puzzle_input[i,"V1"], split = "-"))[1])
  pos2 <- as.integer(unlist(strsplit(puzzle_input[i,"V1"], split = "[- ]"))[2])
  letter <- unlist(strsplit(puzzle_input[i,"V1"], split = "[ ]"))[2]
  pw <- str_trim(puzzle_input[i,"V2"], side = "both")
  
  
  # test each row for conditions
  # test for exactly one condition being true, not both
  # xor() exactly one is true
  if ( xor(substr(pw, pos1, pos1) == letter,substr(pw, pos2,pos2) == letter) ){
    true_ct <- true_ct + 1
  } else {
    false_ct <- false_ct + 1
  }
}

print(true_ct)
print(false_ct)

File: 2020/02/day02.jl
using DelimitedFiles
using DataStructures
puzzle_input = readdlm("/Users/jphilipp/Documents/AoC/2020/02/puzzle_input.txt", '\t', '\n')

## PART 1
function check_valid(x)
    true_ct = 0
    for i in x
       min = parse(Int64,split(i, "-")[1])
    max = parse(Int64,split(split(i,"-")[2])[1])
    letter = only(split(split(i)[2],":")[1])
    pw = split(i,": ")[2]
    ct = counter(pw)
        
        if (ct[letter] >= min && ct[letter] <= max)
        true_ct += 1
        end
    end
    return(true_ct)
end

true_pw = check_valid(puzzle_input)

print("Part 1: Valid passwords: ", true_pw, "\n")



## PART 2
function check_position(x)
    true_ct = 0
    for i in x
        pos1 = parse(Int64,split(i, "-")[1])
        pos2 = parse(Int64,split(split(i,"-")[2])[1])
        letter = only(split(split(i)[2],":")[1])
        pw = split(i,": ")[2]
        
        if (xor(pw[pos1] == letter, pw[pos2] == letter))
            true_ct += 1
        end
    end
    return(true_ct)
end

valid_pw = check_position(puzzle_input)

print("Part 2: Valid passwords: ", valid_pw, "\n")
